### ISSUE TITLE
Getting recursive type errors with TS 4.3.2 when they don't happen with TS 4.2.4

### ISSUE THREAD

**Issue by xaviergonz** [microsoft/TypeScript#44281]
# Bug Report

Given (this is just the smaller repro I could come up with, not real code):
```ts
const p = <T>(fn: () => T): T => {return null as any}

const Base = <T>(val: T): {new(): T} => {
  return null as any;
}

class C extends Base({
 x: p<C[]>(() => [])
}) { }
```

With TS 4.3.2 it gives the errors:
```
Type 'C' recursively references itself as a base type.ts(2310)
'C' is referenced directly or indirectly in its own base expression.ts(2506)
```

where as there's no such error with TS 4.2.4

If the p function parameter is changed from `fn: () => T` to `fn: () => any` or ` x: p<C[]>(() => [])` is changed to ` x: p<C[]>(() => [] as any)` then it does not give that error anymore.

Real impact:

This impacts a library that used to be able to do self-referenced models:
```ts
class TreeNode extends Model({ children: prop<TreeNode[]>(() => []), x: prop(0) }) {}
```
but with the new version it is not possible anymore

Alternative repro:

```ts
abstract class Base2 {
    abstract root() : Derived
}

class Derived extends class extends Base2 {
  root() {
    return undefined as any
  }
}
{}
```

### 🕗 Version & Regression Information

- This changed between versions 4.2.4 and 4.3.2

### ⏯ Playground Link

[Playground Link](https://www.typescriptlang.org/play?jsx=0&ts=4.3.2&strict=true#code/LAKAxg9gdgzgLgAgA4ILwIDwBUB8AKAMygC4E8BKNHBLc0rKhAbwCcBTOAVxagSk4A2AhAEMYoqAE8AvqFCRYiAEJi2aTLjwA3EQPp1mUNgHcK9aYyagECdlx59BwsRMkBuULJDyBY8QGEENgAPODYoABNxFRg2PCsQBGDSJAx-AG0AXXwKRizyT0omBC9QAHoyiQg4AAs2FgQwVTkQEQAjeBYRMEQwXxho1QAmZmsbUQ64Lp7bCGrc0gAReoBLLTYIzxa+vwRlljWNoNDwqMb+8RCwyMHYkYSbFjm4XIfxu25eTki2AhWjCKicQiKRjLxeJjSIA)

**Comment by xaviergonz** [comment]
It seems this was introduced in `4.3.0-dev.20210428` (does not happen in `4.3.0-dev.20210427`).

More specifically it seems to be introduced by the addition of this line 

https://github.com/microsoft/TypeScript/commit/7748694d603c3c604885a6200ce6194a9fa273ef#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R17241

from a commit from @andrewbranch

**Comment by xaviergonz** [comment]
More specifically the call to getBaseTypes on this line seems to trigger some side effect that causes this:

https://github.com/microsoft/TypeScript/commit/7748694d603c3c604885a6200ce6194a9fa273ef#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R20008

**Comment by xaviergonz** [comment]
The error seems to originate in getBaseConstructorTypeOfClass <- resolveBaseTypesOfClass <- getBaseTypes <- getSingleBaseForNonAugmentingSubtype
```
  if (!pushTypeResolution(type, TypeSystemPropertyName.ResolvedBaseConstructorType)) {
    return errorType;
  }
```

I don't think my knowledge is deep enough to dig any deeper :(

**Comment by AndrewMorsillo** [comment]
I think this is what's breaking my mobx-state-tree project when upgrading to TS 4.3. Not sure if this is affecting all MST projects or not but I ended up with a ton of recursive type errors after trying to update to TS 4.3.

**Comment by dsherret** [comment]
Another example showing this bug is in #44284 (I've closed that issue in favour of this one). Just like this issue, it starts happening in 4.3.0-dev.20210428.

**Comment by mrgleba** [comment]
Another repro
```ts
bstract class Base {
    abstract root() : Derived
}

class Derived extends class extends Base {
    root() {
        return undefined as any
    }
}
{   
}
```

**Comment by xaviergonz** [comment]
@DanielRosenwasser

**Comment by DanielRosenwasser** [comment]
Unclear if this is working as intended, but it might be a regression.

**Comment by xaviergonz** [comment]
has it been decided if it is a regression or not?

**Comment by colinhacks** [comment]
Not to pile on, but this has [broken Zod](https://github.com/colinhacks/zod/issues/473) pretty comprehensively. That said, Zod uses some pretty heinous recursive type hacks so I suppose I wouldn't be surprised if this is intentional.

**Comment by vlazh** [comment]
Still happening with 4.3.3 when don't happen with 4.2.4. [Playground](https://www.typescriptlang.org/play?jsx=0&ts=4.4.0-dev.20210616&strict=true#code/PQKgBApgzgNglgOwC4FoAmcoEMBGMJgACSAngA7QDGATnGatPMsAgPYoQC29JKiSEagDMslAiGABYAFChIsfuky58YNigCuUCChwQhrajrT7E4qdJlyAqkjgwoYCTJmkKYAIIISAeRwArCEokMABeMAAlIMM0AB4oJFoEAHMAGjAsbwA+AG4XaTcCLxIAMQ0EYLhWBDCwAAoAOiasamSoAC4M7wBKMKyukjzLAvICAEkhAFEADwEENChYgBV0pdHVgAsIGvCVsEmHAnCECAA3QX7wmcTRJGXV0f6IWe2FtTPBGTAwAH59w6+YE612ot3uYDWFCeL3mjkhEEBfyWWwQgOBhyGrlGYAA0hASFAfEJ4bEPJAYW9in5AsF0gAhdIAZUScGC5LmbxwrFY+EytUSGgglzAADlqiKNDAYCoILFAQBvQHfADaAAUwIgwABrfGsISeAC6nWZtDZzw5jgFCOk32+fwm11eixBYI8aoN9Ky9PSqvSJ3O1CySqBnnd7KdYDpwb+quDnX9giG3wAvsqdSQ9YaZLl8oV9tNKDANCY8QSiSSyeaI1SAkEkN6wCbWSEq7CwFyeRA+eERIdharWVq5TbPKlATNC8XZenMx50qXCcTRqSG03glkg9Ic8M8y7ggvy8vKxTHDWafXI0yWWaT+3ubydmBe9p+4Ph9854CD0uKCvL43ryQTdtyxdwfHoKoEGWYV5TAZIICQTo6l6UJ+iWMBk0xWQQBwiwZGeMhDBCPMAClGR8EVVVoTg4Dsc5agSJJkjAAAfNQNE4PRqFYu9Oz5NiEElGAePKEwhDMNAsIIoiNWQQQRDEMAyIo6k63DNsokoGJ4hZFJ0mUkUADUsCLIUwFg5N8mk6gQn4eTRAIAyPGoUESHUt4oiwNBqhgEhnNc2IDOM0z+gsqzpkImywFI8ijJMwVagMqi4BouiCDYgzVLZDLYv8rBBnyUBwAAGTgHUwAAESwAQwFYWtaTAABZVhOG2EI6vPdIi2maoGiqgQlhSggOrU5xpGs2y5OEBywGCwUAGFqiQLAzGoUl3McJLqNouBzlCwFTniiAiSQlD+g8IZLJ3bEJyLEwygqOxqkWY8LU8bwsqA2pbqnb8K3SYoHsqaoQJGdxgrgNAqNYMgXo297fHqi8fFShd4ZnfVMhIYUfvu8pgYQRZARR2i-xJpA0bYjGlNixlBDgEy4AALxlWIsfSLGN2zLCYoovLfCEd9R0BMYoE87yEF8+GOwfMcR3JtHWzePcKfxF7N2+YVRfFny3KVy1qASv4dclvyXPywKoGqOAhBIMnUbVjcQ35y3rdt+3aIXDceexcCnoQEz1v1hHPvSBW1fhlWF1JJ3wkVEc1Vk-NJxMWJqajtXD1-OcwD9yC2eyL1c4dgksgNH5OhIq2EBtu23VVD1i89x3Lp99wxQQPPqkD17qw+pGw5Lxxg4zgkY5gwFE81amIah6gYedAs7ogP6jzDiDqgL7Gi-D0ujSU6va9Jd1B+b0vW9zbFMqRokg9vM861P1WCUj2ZQX3NXx9qGfGbnhehc-PLIe2Z4YJmoIiTw8M5oQEWsgFaJw1qICEIISIGtfiRDROZZMYBgDAHkEwVAGBsB4B0EwAgxBRhQBoHQBgChmA4EyCgQoUBMFdwDjAD2z8oD9AAGSimqGwnuT8vZtwIKPdWtRe5tjYbEJBKClhoL+OI6CmCpFvAMnTWgjMWYkNkQgZB3E9icwgcohRqj4Ym18i7ORhjFH5huB-MeZiRydDUaefu54IHUw8JgsBUkIoyVIofd2Odd7D1vMojwwowHCjcbnDeUEbEQjsVXN2ds9gq1iGE9IpiNz9DYqJUwJw0DmODho+m2jWZJKMdkCBqSa7uwyW-ME2T7Hvy4dBLmLjIHB0sWbAK1SUm5XNgLcEcSXac3Qc+AgnQrQ5OacELJQ95kOI6Qorp3xXHw2sPwAAHPzHiOzkAAEYABsBy2JHKQAAZgAEwXLAGMZA+yRk8SeUgM5Dz3l3P5hAl2glOIXFKffDxdY6mxU+rfUJQ80FbLKbFaBfzMGMUQMkfxkVJoCGmopcpWj4A6PwHfN6D8GphNfqs6OUTagxPMoCIwXldZgAAPraDxczGUldaYVPxazaFZ9uFJmiqwAyyFK7BLrsIluMgrpAA)

![image](https://user-images.githubusercontent.com/374240/122342963-03f68200-cf67-11eb-81a1-15985196cdc3.png)

**Comment by vlazh** [comment]
Error disappearing if remove
```ts
  : A extends Uint8Array | Uint16Array | Uint32Array | Int8Array | Int16Array | Int32Array
  ? Array<number>
```
but compilation time is about 10x slower than 4.2.4.

**Comment by xaviergonz** [comment]
The reproduction cases are still erroring for me after updating to 4.3.3...

**Comment by xaviergonz** [comment]
Actually the merged fix code is not in TS4.3.3 at all...

**Comment by vlazh** [comment]
Still happening with 4.4.1-rc.

**Comment by vlazh** [comment]
@ahejlsberg, the issue still happening with 4.4.2. Could you reopen the issue?

**Comment by xaviergonz** [comment]
I think you might have better luck if you open a new issue with a playground link to the case that gets broken. as a matter of fact the two reproductions listed in the first post have been fixed in 4.4,so your issue might be because of another reason

**Comment by AndrewMorsillo** [comment]
Yeah something is still wrong in 4.4.2 but I'm not certain if it's this issue. My mobx-state-tree models blow up with thousands of errors if I try to upgrade, same as 4.3. I'm still stuck on 4.2.4 for now.

**Comment by DanielRosenwasser** [comment]
Would you each be able to open up a separate issue with a minimal repro? In the worst-case, we'll dedupe them.

**Comment by vlazh** [comment]
@DanielRosenwasser, separated #45672
